internal/rbd/replicationcontrollerserver.go

```diff
@@ -0,0 +1,357 @@
+		}
+	}
+	return rbdVol, cr, err
+}
+
+// getForceOption extracts the force option from the GRPC request parameters.
+// If not set, the default will be set to false.
+func getForceOption(ctx context.Context, parameters map[string]string) (bool, error) {
+	val, ok := parameters[forceKey]
+	if !ok {
+		util.WarningLog(ctx, "%s is not set in parameters, setting to default (%v)", forceKey, false)
+		return false, nil
+	}
+	force, err := strconv.ParseBool(val)
+	if err != nil {
+		return false, status.Errorf(codes.Internal, err.Error())
+	}
+	return force, nil
+}
+
+// getMirroringMode gets the mirroring mode from the input GRPC request parameters.
+// mirroringMode is the key to check the mode in the parameters.
+func getMirroringMode(ctx context.Context, parameters map[string]string) (librbd.ImageMirrorMode, error) {
+	val, ok := parameters[imageMirroringKey]
+	if !ok {
+		util.WarningLog(ctx, "%s is not set in parameters, setting to mirroringMode to default (%s)", imageMirroringKey, imageMirrorModeSnapshot)
+		return librbd.ImageMirrorModeSnapshot, nil
+	}
+
+	var mirroringMode librbd.ImageMirrorMode
+	switch imageMirroringMode(val) {
+	case imageMirrorModeSnapshot:
+		mirroringMode = librbd.ImageMirrorModeSnapshot
+	default:
+		return mirroringMode, status.Errorf(codes.InvalidArgument, "%s %s not supported", imageMirroringKey, val)
+	}
+	return mirroringMode, nil
+}
+
+// cleanup performs below resource cleanup operations.
+func (cs *ControllerServer) cleanup(rbdVol *rbdVolume, cr *util.Credentials) {
+	if cr != nil {
+		// destroy the credential file
+		cr.DeleteCredentials()
+	}
+	if rbdVol != nil {
+		// release the volume lock
+		cs.VolumeLocks.Release(rbdVol.VolID)
+		// destroy the cluster connection
+		rbdVol.Destroy()
+	}
+}
+
+// EnableVolumeReplication extracts the RBD volume information from the
+// volumeID,If the image is present it will enable the mirroring based on the
+// user provided information.
+// TODO: create new Replication controller struct for the replication operations.
+func (cs *ControllerServer) EnableVolumeReplication(ctx context.Context,
+	req *replication.EnableVolumeReplicationRequest,
+) (*replication.EnableVolumeReplicationResponse, error) {
+	rbdVol, cr, err := cs.getVolumeFromID(ctx, req.GetVolumeId(), req.GetSecrets())
+	defer cs.cleanup(rbdVol, cr)
+	if err != nil {
+		return nil, err
+	}
+	// extract the mirroring mode
+	mirroringMode, err := getMirroringMode(ctx, req.GetParameters())
+	if err != nil {
+		return nil, err
+	}
+
+	mirroringInfo, err := rbdVol.getImageMirroringInfo(ctx)
+	if err != nil {
+		err = fmt.Errorf("failed to get mirroring info of %s with error %s", rbdVol.String(), err)
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if mirroringInfo.State != librbd.MirrorImageEnabled {
+		err = rbdVol.enableImageMirroring(ctx, mirroringMode)
+		if err != nil {
+			util.ErrorLog(ctx, "failed to enable mirroring of %s with error %s", err.Error())
```

Comment:
There is an inconsistency in logging and updateing+returning errors. Soemtimes you gerenera a new `err` with `fmt.Errorf()`, sometimes you log a message with error, and return the original error.

I would opt to extend the error with `%w` in a new error and return it, instead of logging a different message.

Reply:
copy-paste mistake. removed `fmt.Errorf` now just returning the error message what we got.